probe/endpoint/nat.go

```diff
@@ -30,45 +30,102 @@ func makeNATMapper(fw flowWalker) natMapper {
+  we could add 172.31.2.17:8080 (nat original destination) as a copy of ip-172-31-2-17;172.17.0.2:80 (nat reply source)
+
+All of the above can be satisfied by these rules:
+  For SRC_NAT either add NAT orig source as a copy of NAT reply destination
+    or add NAT reply destination as a copy of NAT original source
+  For DST_NAT replace the destination in adjacencies with the NAT reply source
+    and add nat original destination as a copy of nat reply source
+*/
+
+// applyNAT modifies Nodes in the endpoint topology of a report, based on
 // the NAT table.
 func (n natMapper) applyNAT(rpt report.Report, scope string) {
 	n.flowWalker.walkFlows(func(f conntrack.Conn, _ bool) {
-		mapping := toMapping(f)
+		replyDstID := endpointNodeID(scope, f.Reply.Dst, f.Reply.DstPort)
+		origSrcID := endpointNodeID(scope, f.Orig.Src, f.Orig.SrcPort)
 
-		realEndpointPort := strconv.Itoa(int(mapping.originalPort))
-		copyEndpointPort := strconv.Itoa(int(mapping.rewrittenPort))
-		realEndpointID := report.MakeEndpointNodeID(scope, "", mapping.originalIP.String(), realEndpointPort)
-		copyEndpointID := report.MakeEndpointNodeID(scope, "", mapping.rewrittenIP.String(), copyEndpointPort)
-
-		node, ok := rpt.Endpoint.Nodes[realEndpointID]
-		if !ok {
-			return
+		if (f.Status & conntrack.IPS_SRC_NAT) != 0 {
+			if replyDstID != origSrcID {
+				// either add NAT orig source as a copy of NAT reply destination
+				if replyDstNode, ok := rpt.Endpoint.Nodes[replyDstID]; ok {
+					newNode := replyDstNode.WithID(origSrcID).WithLatests(map[string]string{
+						CopyOf: replyDstID,
+					})
+					rpt.Endpoint.AddNode(newNode)
+				} else if origSrcNode, ok := rpt.Endpoint.Nodes[origSrcID]; ok {
+					// or add NAT reply destination as a copy of NAT original source
+					newNode := origSrcNode.WithID(replyDstID).WithLatests(map[string]string{
+						CopyOf: origSrcID,
+					})
+					rpt.Endpoint.AddNode(newNode)
+				}
+			}
 		}
 
-		rpt.Endpoint.AddNode(node.WithID(copyEndpointID).WithLatests(map[string]string{
-			CopyOf: realEndpointID,
-		}))
+		if (f.Status & conntrack.IPS_DST_NAT) != 0 {
+			fromID := endpointNodeID(scope, f.Reply.Dst, f.Reply.DstPort)
+			fromNode, ok := rpt.Endpoint.Nodes[fromID]
+			if !ok {
+				return
+			}
+			toID := endpointNodeID(scope, f.Orig.Dst, f.Orig.DstPort)
+
+			// replace destination with reply source
+			replySrcID := endpointNodeID(scope, f.Reply.Src, f.Reply.SrcPort)
+			if replySrcID != toID {
+				fromNode.Adjacency = fromNode.Adjacency.Minus(toID)
+				fromNode = fromNode.WithAdjacent(replySrcID)
+
+				// add nat original destination as a copy of nat reply source
+				replySrcNode, ok := rpt.Endpoint.Nodes[replySrcID]
+				if !ok {
+					replySrcNode = report.MakeNode(replySrcID)
+				}
+				newNode := replySrcNode.WithID(toID).WithLatests(map[string]string{
+					CopyOf: replySrcID,
+				})
+				rpt.Endpoint.AddNode(newNode)
+			}
+
+		}
```

Comment:
The difference in structure between the two 'big ifs' has me scratching my head. Could we regularise this to
```golang
if (f.Status & conntrack.IPS_SRC_NAT) != 0 {
    replyDstID := endpointNodeID(scope, f.Reply.Dst, f.Reply.DstPort)
    origSrcID := endpointNodeID(scope, f.Orig.Src, f.Orig.SrcPort)
    if replyDstID != origSrcID { ... }
}
if (f.Status & conntrack.IPS_DST_NAT) != 0 {
    replySrcID := endpointNodeID(scope, f.Reply.Src, f.Reply.SrcPort)
    origDstID := endpointNodeID(scope, f.Orig.Dst, f.Orig.DstPort)
    if replySrcID != origDstID { ... }
}
```
?